### PR TITLE
Get async test function name

### LINKED ISSUE #15: Fail to get test function name with `test() async throws`
Hi, Thank you a lot for having this project; I love it. 

I found an `ERROR` when using `verify()`

With the following test update, you can see the file name become `*.ERROR.*`

> The approved file name is `ApprovalTests_SwiftTests/Namer/NamerTests.ERROR.story1.approved.md`

```diff
diff --git a/ApprovalTests_SwiftTests/Namer/NamerTests.ERROR.story1.approved.md b/ApprovalTests_SwiftTests/Namer/NamerTests.ERROR.story1.approved.md
new file mode 100644
index 0000000..94f111a
--- /dev/null
+++ b/ApprovalTests_SwiftTests/Namer/NamerTests.ERROR.story1.approved.md
@@ -0,0 +1,2 @@
+# Title
+> Quoted text
diff --git a/ApprovalTests_SwiftTests/Namer/NamerTests.swift b/ApprovalTests_SwiftTests/Namer/NamerTests.swift
index fc5171e..d56e0d7 100644
--- a/ApprovalTests_SwiftTests/Namer/NamerTests.swift
+++ b/ApprovalTests_SwiftTests/Namer/NamerTests.swift
@@ -9,6 +9,9 @@ final class NamerTests: XCTestCase {
     func test_parameters() throws {
         try Approvals.verify("# Title\n> Quoted text", Approvals.Names.withParameters("story1").forFile.with(extensionWithDot: ".md"))
     }
+    func test_parameters_async() async throws {
+        try Approvals.verify("# Title\n> Quoted text", Approvals.Names.withParameters("story1").forFile.with(extensionWithDot:  ".md"))
+    }
 
     func test_multipleParameters() throws {
         let foo = Approvals.Names.withParameters("story1")

```

I would love to contribute, but I am confused about why check the process for the test function name. There is a way to inject function seamlessly, just like file and line. 

```diff
diff --git a/ApprovalTests.Swift/Approvals.swift b/ApprovalTests.Swift/Approvals.swift
index bdc75bf..5ab5978 100644
--- a/ApprovalTests.Swift/Approvals.swift
+++ b/ApprovalTests.Swift/Approvals.swift
@@ -29,6 +29,7 @@ public enum Approvals {
         _ object: T,
         _ options: Options = Options(),
         file: StaticString = #filePath,
+        function: StaticString = #function,
         line: UInt = #line
     ) throws {
         let description = String(describing: object)

```

### PR DESCRIPTION
**Pull request by the-yuyut**
## Description

Get the test case name from the function instead of the call stack.

fix #15

### CODE REVIEW

File: ApprovalTests.Swift/Namer/Namer.swift [hunk 1]
```diff
@@ -13,9 +15,8 @@ public class Namer: ApprovalNamer {
     }
 
     public func sourceFilePath() -> String {
-        let names = StackDemangler().extractNames()
         let baseName = (fileName as NSString).deletingPathExtension
-        return baseName + "." + names.testName
+      return baseName + "." + "\(functionName.dropLast(2))"
```
Comment (benlenarts) [inline ApprovalTests.Swift/Namer/Namer.swift:19]: What does `.dropLast(2)` achieve? You could extract this a variable to describe the result.

Also for formatting:
```suggestion
        return baseName + "." + "\(functionName.dropLast(2))"
```

File: ApprovalTests.Swift/Core/Options.swift [hunk 2]
```diff
@@ -63,7 +63,7 @@ public struct Options {
         data[.scrubber] as? Scrubber ?? ScrubNothing()
     }
 
-    public func with(namerCreator: @escaping (String) -> ApprovalNamer) -> Options {
+    public func with(namerCreator: @escaping (String, String) -> ApprovalNamer) -> Options {
```
Comment (benlenarts) [inline ApprovalTests.Swift/Core/Options.swift:66]: What are the arguments of the closure here? You could name them to make this more obvious. Perhaps even create a type alias for the closure.

File: ApprovalTests.Swift/Namer/ParameterizedNamer.swift [hunk 3]
```diff
@@ -1,9 +1,9 @@
 class ParameterizedNamer: Namer {
     private let parameters: [String]
 
-    public init(_ filePath: String, _ parameters: [String]) {
+  public init(_ filePath: String, _ function: String, _ parameters: [String]) {
```
Comment (benlenarts) [inline ApprovalTests.Swift/Namer/ParameterizedNamer.swift:4]: Formatting:
```suggestion
    public init(_ filePath: String, _ function: String, _ parameters: [String]) {
```

File: ApprovalTests_SwiftTests/ApprovalTests_SwiftTests.swift [hunk 4]
```diff
@@ -7,12 +7,12 @@ import XCTest
 
 final class ApprovalTests_SwiftTests: XCTestCase {
     func testApprovalName() throws {
-        let name = Approvals.makeNamer(forFile: "DUMMY")
+      let name = Approvals.makeNamer(forFile: "DUMMY", function: "testApprovalName")
```
Comment (benlenarts) [inline ApprovalTests_SwiftTests/ApprovalTests_SwiftTests.swift:10]:
```suggestion
        let name = Approvals.makeNamer(forFile: "DUMMY", function: "testApprovalName")
```

File: ApprovalTests_SwiftTests/ApprovalTests_SwiftTests.swift [hunk 5]
```diff
@@ -7,12 +7,12 @@ import XCTest
...
         XCTAssertEqual(name.approvalName(), "ApprovalTests_SwiftTests.testApprovalName")
     }
 
     func testSourceFilePath() throws {
-        let name = Approvals.makeNamer(forFile: #file)
+      let name = Approvals.makeNamer(forFile: #file, function: #function)
```
Comment (benlenarts) [inline ApprovalTests_SwiftTests/ApprovalTests_SwiftTests.swift:15]:
```suggestion
        let name = Approvals.makeNamer(forFile: #file, function: #function)
```

File: ApprovalTests_SwiftTests/Namer/NamerTests.swift [hunk 6]
```diff
@@ -10,10 +10,14 @@ final class NamerTests: XCTestCase {
...
 
+  func test_parameters_async() async throws {
+        try Approvals.verify("# Title\n> Quoted text", Approvals.Names.withParameters("story1").forFile.with(extensionWithDot: ".md"))
+    }
+
     func test_multipleParameters() throws {
         let foo = Approvals.Names.withParameters("story1")
                     .and(Approvals.Names.withAdditionalParameters("part1"))
-        let name: String = foo.getNamer("").sourceFilePath()
+      let name: String = foo.getNamer("", #function).sourceFilePath()
```
Comment (benlenarts) [inline ApprovalTests_SwiftTests/Namer/NamerTests.swift:20]:
```suggestion
        let name: String = foo.getNamer("", #function).sourceFilePath()
```

File: ApprovalTests.Swift/Approvals.swift [hunk 7]
```diff
@@ -109,9 +116,10 @@ public enum Approvals {
         _ object: T,
         _ options: Options = Options(),
         file: StaticString = #filePath,
+        function: StaticString = #function,
         line: UInt = #line
     ) throws {
-        try verify(StringUtils.toJSON(object), options.forFile.with(extensionWithDot: ".json"), file: file, line: line)
+      try verify(StringUtils.toJSON(object), options.forFile.with(extensionWithDot: ".json"), file: file, function: function, line: line)
```
Comment (benlenarts) [inline ApprovalTests.Swift/Approvals.swift:122]:
```suggestion
        try verify(StringUtils.toJSON(object), options.forFile.with(extensionWithDot: ".json"), file: file, function: function, line: line)
```

File: ApprovalTests_SwiftTests/Namer/NamerTests.swift [hunk 8]
```diff
@@ -10,10 +10,14 @@ final class NamerTests: XCTestCase {
         try Approvals.verify("# Title\n> Quoted text", Approvals.Names.withParameters("story1").forFile.with(extensionWithDot: ".md"))
     }
 
+  func test_parameters_async() async throws {
```
Comment (benlenarts) [inline ApprovalTests_SwiftTests/Namer/NamerTests.swift:13]:
```suggestion
    func test_parameters_async() async throws {
```

**Review by benlenarts — COMMENTED** [review]
Love this change! Straightforward and succinct. 👌 

I also ran into not being able to use this package with async test methods.

I spotted a few indentation issues and added a couple questions with suggestions. Apart from those, it looks solid!

**Comment by babbage** [comment]
This is an important PR for me. Without this PR, all async tests in a test class generate the same filename of [TestClassName].ERROR.approved.txt so different tests are trying to verify against the same source. This obviously doesn't work.

Seems like a straightforward and it would be good to see it merged. :)

File: ApprovalTests.Swift/Namer/Namer.swift (hunk 1 above)
Comment (jonreid) [reply ApprovalTests.Swift/Namer/Namer.swift:19]: Thank you for pointing this out. We extracted an explaining variable in 30eba989f6b3b1e9ceaedd5acf20cce131e2ff1e

**Comment by jonreid** [comment]
Thank you @ytyubox! Sorry it took to long for me & Llewellyn to review it together.

File: ApprovalTests.Swift/Core/Options.swift (hunk 2 above)
Comment (jonreid) [reply ApprovalTests.Swift/Core/Options.swift:66]: This reveals a deeper need (beyond this change) for a missing type.

**Pull request merged**